proposals/0000-std-experimental.rst

```diff
@@ -0,0 +1,177 @@
+it might change in future releases. ::
+
+  -- Without -experimental
+  ghc Main.hs  -- This will throw an error if Feature X is used
+
+  -- With -experimental
+  ghc -experimental Main.hs  -- This will compile successfully with Feature X
+
+  -- With -no-experimental
+  ghc -experimental -no-experimental Main.hs -- This will throw an error if Feature X is used
+
+  ghc -no-experimental Main.hs -package foo   -- This will throw an error if Feature X is used anywhere in Main.hs or the package foo.
+
+Effect and Interactions
+-----------------------
+The introduction of this flag provides a clear distinction between stable and
+experimental features. It ensures that developers are aware of the potential
+risks of using experimental features while providing a sandbox for testing out
+new GHC capabilities.
+
+Costs and Drawbacks
+-------------------
+- There might be an initial learning curve for developers to understand the
+  distinction between experimental and stable features.
+- Maintaining two sets of features (stable and experimental) might increase the
+  complexity of the compiler codebase.
+
+The authors believe the benefit of clearly separating experimental and those
+that adhere to a more rigourous change evolution easily offsets the costs for
+both end users and feature developers. For developers of experimental feature
+this also means they reserve the rigth to break them without warning in
+backwards incompatible ways at any moment in time leading to significantly
+lighter processes for experimental features.
+
+Prior Work
+----------
+The ``-experimental`` feature is similar to solutions for experimental
+features and implementation in other languages, and GHC would not have a
+bespoke solution here, but share a common approach with many other compilers.
+Do note that the provided approaches mix editions, channels, and feature gates.
+The goal is to illustrate that the concept of separating experimental features
+explicitly behind opt-in gates.
+
+- **Rust**: Rust has so called channels, which delineate stability by stable,
+  beta, and nightly. To opt-in locally to nightly features you'd use the
+  ``+nightly`` toolchain modifier. Effectively using a different Rust compiler
+  provided by ``rustup``.
+- **Java**: Java has the ``--enable-preview`` flag since Java 9, which allows
+  opt-in to experimental features.
+- **Gnu Compiler Collection**: GCC provides flags to opt-in to experimental
+  features through the ``-std=...`` flag. For example, it has flags like
+  ``-std=c++1z`` for experimental C++ features before they were standardized in
+  C++17.
+
+Backward Compatibility
+----------------------
+The introduction of this flag is expected to have minimal impact on existing
+code. The initial set of features behind the ``-experimental`` flag will be
+the emtpy set. Follow up proposals will be required to move existing features
+behind the `-experimental` flag. New features should by default be behind
```

Comment:
Isn't moving a feature behind `-experimental` a breaking change? :sweat_smile:

Reply:
The beauty of the proposal is that stable part is not actually stable. it just won't change without a deprecation period. So we can still argue on how long something needs to be deprecated before moved to `-experimental`. And then we can argue whether the feature (after seeing no changes being moved to or initially implemented in `-experimental`) is stable! /s

Reply:
Despite @phadej `/s`, yes, you'd be _deprecating_ existing experimental features into `-experimental`. Tell people what's going to happen and then make it happen.

For _extensions_ there is #601 which aims to describe a lifecycle. This would cover features as well. But most importantly, I think what is in -experimental and not in -experimental, should primarily be governed by how we treat it. You can't make breaking changes without an change evolution in -no-experimental parts of the compiler, while you can do that pretty freely with -experimental features and extensions.